### PR TITLE
Explosion Performance Upgrades

### PR DESCRIPTION
**Pull request by RemieRichards**
Before:
## DEBUG: Explosion(113,135,1)(d3,h7,l14): Took 3.2 seconds.

After:
## DEBUG: Explosion(113,135,1)(d3,h7,l14): Took 1 seconds.

(Tested on box, at the bridge, where the AI holopad is, both bombs)

Thanks to @clusterfack for discussing the topic of datum references being inefficient.
- subsystem `postpone()` now takes an optional `cycles` argument, to allow for postponing for more than one cycle
- powernet datums now manually throw away references, because byond sucks.
- adds a sanity check to item throwing, because objects were apparently deleting between `spawn(0)`

### CODE REVIEW

**Comment by Iamgoofball** [comment]
Bombcap increase when :+1:

**Comment by xxalpha** [comment]
>optimize explosions so there's less lag
>increase bombcap so there's more lag

**Comment by WJohn** [comment]
It's okay.

It all balances out in the end.

Optimizations are purely there to increase the limits, and by god we will push them.

File: code/game/objects/explosion.dm [hunk 1]
```diff
@@ -112,10 +114,11 @@ proc/explosion(turf/epicenter, devastation_range, heavy_impact_range, light_impa
 			var/throw_dir = get_dir(epicenter,T)
 			for(var/obj/item/I in T)
 				spawn(0) //Simultaneously not one at a time
```
Comment (Miauw62) [inline code/game/objects/explosion.dm]: TBH this spawn() probably causes a bunch of lag too, and I don't think it actually makes that much difference.

**Comment by Miauw62** [comment]
>21 lines of code changed
>150% increase in performance

:')

**Comment by clusterfack** [comment]
Sometimes a little knowledge goes a long way

File: code/game/objects/explosion.dm (hunk 1 above)
Comment (RemieRichards) [inline code/game/objects/explosion.dm]: @Miauw62 it's a visual effect, if you read the comment, it causes each thrown item to be thrown at once (you know, as if they'd all been pushed by a big force, like an explosion)

Without it, they move 1 and a time, which is horrific visually.

Comment (MrStonedOne) [inline code/game/objects/explosion.dm]: throw_at uses sleep in it, so ya, you need the spawn, other wise things would wait until the last item reached its destination before moving on to the next

**Comment by optimumtact** [comment]
What was the outcome of that discussion if you don't mind?

**Comment by clusterfack** [comment]
Byond cannot efficiently del references of any kind. And datum deletion is profiled under the self cpu of the proc where it is called, instead of being profiled as its own proc. So by manually removing references in destroy you massively decrease the amount of cpu wasted in deleting datums.

Also powernets on most servers forked directly from TG rebuild the powernet when a wire is cut, in an explosion it will rebuild a powernet then cut another wire and the rebuild it again, and you can often see it happen 100+ times, the other alternatives primarily being deferring the powernet rebuilding during explosions.

**Comment by RemieRichards** [comment]
@optimumtact yeah what @clusterfack said, they got to it before me :P 
tl;dr:
* remove references on big datums manually for a perfomance boost
* Datum destruction profiling is oddly placed in the proc that destroyed the datum, not the datum's own destruction profile.

**Comment by optimumtact** [comment]
Thanks guys
